packages/component-base/src/data-provider-controller.js

```diff
@@ -0,0 +1,118 @@
+import { Cache } from './data-provider-controller/cache.js';
+import { getFlatIndexByPath, getFlatIndexInfo } from './data-provider-controller/helpers.js';
+
+export class DataProviderController extends EventTarget {
+  constructor(host, { size, pageSize, isExpanded, dataProvider, dataProviderParams }) {
+    super();
+    this.host = host;
+    this.size = size;
+    this.pageSize = pageSize;
+    this.isExpanded = isExpanded;
+    this.dataProvider = dataProvider;
+    this.dataProviderParams = dataProviderParams;
+    this.rootCache = this.#createRootCache();
+  }
+
+  get effectiveSize() {
+    return this.rootCache.effectiveSize;
+  }
+
+  isLoading() {
+    return this.rootCache.isLoading;
+  }
+
+  setSize(size) {
+    const delta = size - this.rootCache.size;
+    this.size = size;
+    this.rootCache.size += delta;
+    this.rootCache.effectiveSize += delta;
```

Comment:
Having to have to manually update the cache's effective size somehow feels dangerous because it means the cache can end up in an invalid state just by using its public API (if you only update the size). I think an earlier revision of this PR had `size` as a setter that automatically handled the update of effective size also?

Reply:
This is actually a direct copy-paste from the original implementation where it was apparently an optimization to avoid traversing the entire cache tree when updating the root cache size. 

I agree that it can be dangerous. There is a general problem that the cache can end up in an invalid state when decreasing the size while some items are expanded since the cache doesn't know which items exactly the size decrease will affect. In that case, it would probably be more correct to clear the cache. This is, however, a problem that exists in the original implementation. 

Should we consider calling `this.recalculateEffectiveSize()` for consistency here?